// dynamic-presentation/js/AppleMusicPlayer.js
/* globals jQuery */

import { VoidPlayer } from '../js/VoidPlayer.js?v=1.8.0';

export class AppleMusicPlayer extends VoidPlayer {
  constructor (url, onUpdate, onStop, onEnd, playbackRate) {
    super(url, onUpdate, onStop, onEnd, playbackRate);
    let bodyTag = document.getElementsByTagName('body')[0];
    this._appleMusicCode = this._extractVideoID(url);
    this._appleMusicUrl =
      'https://embed.music.apple.com/us/album/' + this._appleMusicCode;
    this._iframeText = `
      <iframe
        id="apple-music-player"
        width="400"
        height="163"
        src=""
        frameborder="0"
        allow="encrypted-media *; fullscreen *"
        style="position: absolute; opacity: 0.5; z-index: 20; overflow:hidden; background:transparent;"
        sandbox="allow-forms allow-popups allow-same-origin allow-scripts allow-top-navigation-by-user-activation"
      ></iframe>
    `;
    bodyTag.insertAdjacentHTML('afterbegin', this._iframeText);
    jQuery('#apple-music-player').hide();
  }

  hide () {
    jQuery('#apple-music-player').attr('src', '');
    jQuery('#apple-music-player').hide();
  }

  show () {
    jQuery('#youtube-player').hide();
    jQuery('#bandcamp-player').hide();
    jQuery('#spotify-player').hide();
    if (jQuery('#apple-music-player').attr('src') !== this._appleMusicUrl) {
      jQuery('#apple-music-player').attr('src', this._appleMusicUrl);
    }
    jQuery('#apple-music-player').show();
  }

  // Until Apple Music has an API to control the player, this doesn't do anything
  play () {
    this._playing = true;
    this._update(true);
  }
  pause () {
    this._playing = false;
  }
  stop () {
    jQuery('#apple-music-player').hide();
    this.onEnd();
    this.onStop();
  }
  getTimestamp () {
    return 0;
  }

  _extractVideoID (url) {
    var regExp = /music\.apple\.com\/[a-zA-Z]*\/album\/.*?([0-9]+\?i=[0-9]+)/;
    var match = url.match(regExp);
    if (match && match[1].length) {
      return match[1];
    }
    console.log('Could not extract video ID.');
    return false;
  }
}
