// dynamic-presentation/js/BandcampPlayer.js
/* globals jQuery */

import { VoidPlayer } from '../js/VoidPlayer.js?v=1.8.0';

export class BandcampPlayer extends VoidPlayer {
  constructor (url, onUpdate, onStop, onEnd, playbackRate) {
    super(url, onUpdate, onStop, onEnd, playbackRate);
    let bodyTag = document.getElementsByTagName('body')[0];
    this._bandcampCode = this._extractVideoID(url);
    this._bandcampUrl =
      'https://bandcamp.com/EmbeddedPlayer/' +
      this._bandcampCode +
      '/size=large/bgcol=ffffff/linkcol=63b2cc/tracklist=false/artwork=small/transparent=true/';
    this._iframeText = `
      <iframe
        id="bandcamp-player"
        width="400"
        height="120"
        src=""
        frameborder="0"
        seamless
        style="position: absolute; opacity: 0.5; z-index: 20;"
      ></iframe>
    `;
    bodyTag.insertAdjacentHTML('afterbegin', this._iframeText);
    jQuery('#bandcamp-player').hide();
  }

  hide () {
    jQuery('#bandcamp-player').attr('src', '');
    jQuery('#bandcamp-player').hide();
  }

  show () {
    jQuery('#youtube-player').hide();
    jQuery('#spotify-player').hide();
    jQuery('#apple-music-player').hide();
    if (jQuery('#bandcamp-player').attr('src') !== this._bandcampUrl) {
      jQuery('#bandcamp-player').attr('src', this._bandcampUrl);
    }
    jQuery('#bandcamp-player').show();
  }

  // Until Bandcamp has an API to control the player, this doesn't do anything
  play () {
    this._playing = true;
    this._update(true);
  }
  pause () {
    this._playing = false;
  }
  stop () {
    jQuery('#bandcamp-player').hide();
    this.onEnd();
    this.onStop();
  }
  getTimestamp () {
    return 0;
  }

  _extractVideoID (url) {
    var regExp =
      /bandcamp\.com\/EmbeddedPlayer.*\/album=([0-9]+).*\/track=([0-9]+)/;
    var match = url.match(regExp);
    if (match && match[1].length && match[2].length) {
      return 'album=' + match[1] + '/track=' + match[2];
    }
    console.log('Could not extract video ID.');
    return false;
  }
}
