// dynamic-presentation/js/SpotifyPlayer.js
/* globals jQuery */

import { VoidPlayer } from '../js/VoidPlayer.js?v=1.8.0';

export class SpotifyPlayer extends VoidPlayer {
  constructor (url, onUpdate, onStop, onEnd, playbackRate) {
    super(url, onUpdate, onStop, onEnd, playbackRate);
    let bodyTag = document.getElementsByTagName('body')[0];
    this._spotifyCode = this._extractVideoID(url);
    this._spotifyUrl =
      'https://open.spotify.com/embed/track/' +
      this._spotifyCode +
      '?utm_source=generator';
    this._iframeText = `
      <iframe
        id="spotify-player"
        width="400"
        height="80"
        src=""
        frameBorder="0"
        allowfullscreen=""
        allow="clipboard-write; encrypted-media; fullscreen; picture-in-picture"
        style="position: absolute; opacity: 0.5; z-index: 20;"
      ></iframe>
    `;
    bodyTag.insertAdjacentHTML('afterbegin', this._iframeText);
    jQuery('#spotify-player').hide();
  }

  hide () {
    jQuery('#spotify-player').attr('src', '');
    jQuery('#spotify-player').hide();
  }

  show () {
    jQuery('#youtube-player').hide();
    jQuery('#bandcamp-player').hide();
    jQuery('#apple-music-player').hide();
    if (jQuery('#spotify-player').attr('src') !== this._spotifyUrl) {
      jQuery('#spotify-player').attr('src', this._spotifyUrl);
    }
    jQuery('#spotify-player').show();
  }

  // Until Spotify has an API to control the player, this doesn't do anything
  play () {
    this._playing = true;
    this._update(true);
  }
  pause () {
    this._playing = false;
  }
  stop () {
    jQuery('#spotify-player').hide();
    this.onEnd();
    this.onStop();
  }
  getTimestamp () {
    return 0;
  }

  _extractVideoID (url) {
    var regExp = /open\.spotify\.com\/track\/([a-zA-Z0-9]*)/;
    var match = url.match(regExp);
    if (match && match[1].length === 22) {
      return match[1];
    }
    console.log('Could not extract video ID.');
    return false;
  }
}
